packages/dynamic_layouts/lib/src/staggered_layout.dart

```diff
@@ -0,0 +1,279 @@
+  final Axis scrollDirection;
+
+  /// The collection of scroll offsets for every row or column across the main
+  /// axis. It includes the tiles' sizes and the spacing between them.
+  final Map<int, double> _scrollOffsetForMainAxis = <int, double>{};
+
+  /// The amount of tiles in every row or column across the main axis.
+  final Map<int, int> _mainAxisCount = <int, int>{};
+
+  /// Returns the row or column with the minimum extent for the next child to
+  /// be layed out.
+  int _getNextCrossAxisSlot() {
+    int nextCrossAxisSlot = 0;
+    double minScrollOffset = double.infinity;
+    _scrollOffsetForMainAxis.forEach((int mainAxisSeries, double scrollOffset) {
+      if (scrollOffset < minScrollOffset) {
+        nextCrossAxisSlot = mainAxisSeries;
+        minScrollOffset = scrollOffset;
+      }
+    });
+    return nextCrossAxisSlot;
+  }
+
+  /// Returns true if every child has crossed the [targetOffset] and false
+  /// if not.
+  @override
+  bool reachedTargetScrollOffset(double targetOffset) {
+    for (final double scrollOffset in _scrollOffsetForMainAxis.values) {
+      if (scrollOffset < targetOffset) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /// The infinite extent in the main axis lets
+  /// [RenderDynamicSliverGrid.performLayout()] know that the child will decide
+  /// its own size in that axis.
+  @override
+  DynamicSliverGridGeometry getGeometryForChildIndex(int index) {
+    return DynamicSliverGridGeometry(
+      scrollOffset: 0.0,
+      crossAxisOffset: 0.0,
+      mainAxisExtent: double.infinity,
+      crossAxisExtent: childCrossAxisExtent,
+    );
+  }
+
+  /// Returns an updated [DynamicSliverGridGeometry] with the actual size of
+  /// the child.
+  @override
+  DynamicSliverGridGeometry updateGeometryForChildIndex(
+      int index, Size childSize) {
+    final int crossAxisSlot = _getNextCrossAxisSlot();
+    final double currentScrollOffset = _scrollOffsetForMainAxis[crossAxisSlot]!;
+    final double childMainAxisExtent =
+        scrollDirection == Axis.vertical ? childSize.height : childSize.width;
+    _scrollOffsetForMainAxis[crossAxisSlot] =
+        childMainAxisExtent + _scrollOffsetForMainAxis[crossAxisSlot]!;
+
+    final DynamicSliverGridGeometry sliverGridGeometry =
+        DynamicSliverGridGeometry(
+      scrollOffset: currentScrollOffset +
+          _mainAxisCount[crossAxisSlot]! * mainAxisSpacing,
+      crossAxisOffset:
+          crossAxisSlot * (childCrossAxisExtent + crossAxisSpacing),
+      mainAxisExtent:
+          scrollDirection == Axis.vertical ? childSize.height : childSize.width,
+      crossAxisExtent: childCrossAxisExtent,
+    );
+    _mainAxisCount[crossAxisSlot] = _mainAxisCount[crossAxisSlot]! + 1;
+    return sliverGridGeometry;
+  }
+}
+
+/// Creates dynamic grid layouts with a fixed number of tiles in the cross axis
+/// and varying main axis size. It uses the same logic as
+/// [SliverGridDelegateWithFixedCrossAxisCount], but with a
```

Comment:
The user may not know what the logic is, can you explain more? Some sample code would be great too.